pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/BalancedInstanceSelector.java

```diff
@@ -89,7 +91,13 @@ Pair<Map<String, String>, Map<String, String>> select(List<String> segments, int
         if (candidates == null) {
           continue;
         }
-        int selectedIdx = requestId++ % candidates.size();
+        int selectedIdx;
+        if (_useStickyRouting || QueryOptionsUtils.isUseStickyRouting(queryOptions)) {
+          // candidates array is always sorted
+          selectedIdx = Math.abs(_tableNameWithType.hashCode() % candidates.size());
```

Comment:
Consider using raw table name so that we can pick the same replica for hybrid use cases for future proof. Also we can cache the hash (positive value) to avoid calculating hash per query

Reply:
Ack